app/queues/notifications/expiration-notification.server.ts

```diff
@@ -0,0 +1,152 @@
+  DnsRecord = 'record',
+}
+export interface ExpirationStatusPayload {
+  type: RecordType;
+}
+
+// name for the queue
+const expirationNotificationQueueName = 'expiration-notification';
+
+const updateNotificationStatus = (type: RecordType, id: number) => {
+  switch (type) {
+    case RecordType.Certificate:
+      return prisma.certificate.update({
+        where: {
+          id,
+        },
+        data: {
+          lastNotified: new Date(),
+        },
+      });
+    case RecordType.DnsRecord:
+      return prisma.record.update({
+        where: {
+          id,
+        },
+        data: {
+          lastNotified: new Date(),
+        },
+      });
+  }
+};
+
+// fetch records with by status
+const getRecordsByExpiration = (type: RecordType) => {
+  let monthFromToday = new Date();
+  monthFromToday.setMonth(monthFromToday.getMonth() + 1);
+
+  switch (type) {
+    case RecordType.Certificate:
+      return prisma.certificate.findMany({
+        where: {
+          validTo: {
+            lte: monthFromToday,
+          },
+          lastNotified: null,
+        },
+        select: {
+          user: true,
+          id: true,
+          lastNotified: true,
+        },
+      });
+    case RecordType.DnsRecord:
+      return prisma.record.findMany({
+        where: {
+          expiresAt: {
+            lte: monthFromToday,
+          },
+          lastNotified: null,
+        },
+        select: {
+          user: true,
+          id: true,
+          lastNotified: true,
+        },
+      });
+  }
+};
+// Queue initialization
+export const expirationNotificationQueue = new Queue<ExpirationStatusPayload>(
+  expirationNotificationQueueName,
+  {
+    connection: redis,
+  }
+);
+
+expirationNotificationQueue.on('error', (err) => {
+  logger.warn('Error running check for DNS Records/Certificates about to expire', err);
+});
+
+// function to add jobs
+export const addExpirationNotification = async (type: RecordType) => {
+  let jobName = `${expirationNotificationQueueName}-${type}`;
+  return expirationNotificationQueue.add(
+    jobName,
+    { type },
+    {
+      repeat: { every: 5 * 60 * 1000 },
+    }
+  );
+};
+// function to update notification and add notification jobs
+const updateStatusAndNotify = async (type: RecordType, id: number, data: NotificationData) => {
+  const { emailAddress, subject, message } = data;
+  await updateNotificationStatus(type, id);
+  await addNotification({
+    emailAddress,
+    subject,
+    message,
+  });
+};
+// worker instance to process DNS Record/Certificates expiration notification jobs
+const expirationNotificationWorker = new Worker<ExpirationStatusPayload>(
+  expirationNotificationQueueName,
+  async (job) => {
+    const { type } = job.data;
+    try {
+      logger.debug(`Processing job: ${job.name}`);
+      let records = await getRecordsByExpiration(type);
+      await Promise.all(
+        records.map(async (record) => {
+          if (!record.lastNotified) {
```

Comment:
We're going to need something more here, so we can deal with renewals.  The first time we notify a user for a record/cert, it will work.  But if they renew, it won't ever notify again in the future.  I wonder if this should really be a check to see if it's a) not null/undefined; or b) a date more than 30 days in the past?

Reply:
> We're going to need something more here, so we can deal with renewals. The first time we notify a user for a record/cert, it will work. But if they renew, it won't ever notify again in the future. I wonder if this should really be a check to see if it's a) not null/undefined; or b) a date more than 30 days in the past?

We could do a combination of both. Initially, it's null, and it is our cue to tell that we have never notified a user. We should also check whether the value of `lastNotified` is set to a month in the past (as you are suggesting), as this would be our cue to re-notify the user.